Visualize PropertyLayers

EwoutH opened the pull request:
This PR adds support for visualizing PropertyLayers in the Matplotlib-based space visualization component. It allows users to overlay PropertyLayer data on top of the existing grid and agent visualizations, or on its own.

It introduces a new `propertylayer_portrayal` parameter to customize the appearance of PropertyLayers and refactors the existing space visualization code for better modularity and flexibility.

### Motive
The PropertyLayer is becoming a key feature in Mesa, so it should be visualized easily. This feature enables users to easily visualize multiple layers of data simultaneously, such as environmental factors or agent properties, alongside the agents themselves.

### Implementation
- Added a new `propertylayer_portrayal` parameter to the `make_space_matplotlib` function.
- Implemented a `draw_property_layers` function to render PropertyLayer data using Matplotlib's `imshow`.
- Modified the `_draw_grid` function to incorporate PropertyLayer visualization before drawing agents.
- Added support for both color-based and colormap-based PropertyLayer rendering.
- Implemented alpha blending for PropertyLayer visualization to allow overlaying multiple layers.
- Added colorbar support for PropertyLayer visualization.

### Usage Examples
You can visualize multiple PropertyLayers ("grass" and "temperature" in this case) along with agents in a grid space:
```python
def agent_portrayal(agent):
    return {"color": "red", "shape": "o", "size": 20}

propertylayer_portrayal = {
    "grass": {
        "colormap": "Greens",
        "alpha": 1,
    },
    "temperature": {
        "colormap": "coolwarm",
        "alpha": 0.333,
        "vmin": 0,
        "vmax": 100
    },
}

model = ExampleModel(10, 10)
page = SolaraViz(
    model,
    [make_space_matplotlib(agent_portrayal, propertylayer_portrayal)],
)
```
![image](https://github.com/user-attachments/assets/45a7eab4-3225-451b-aa98-6fd5702c1c22)

You can also use a single color for a PropertyLayer. Here we render the "cell_layer" PropertyLayer using shades of black, with the highest value having 0% opacity and the lowest value having 100% opacity. For Game of Life that means alive cells are fully black, and dead cells transparent.

```python
propertylayer_portrayal = {
    "cell_layer": {
        "color": "Black",
        "alpha": 1,
        "colorbar": False,
    },
}
```
![image](https://github.com/user-attachments/assets/b8b576be-219e-4711-8dc8-112631947093)

Any [matplotlib color](https://matplotlib.org/stable/api/colors_api.html#module-matplotlib.colors) is valid. The highest value will have that color with 0% opacity, the lowest with 100% opacity (which can again be scaled with modifying `vmin` and `vmax`).

## Additional Notes
- Currently supported spaces: Grid and standalone PropertyLayers.
  - ContinuousSpace, NetworkGrid, and VoronoiGrid can be added in future PR(s).
- Future work could include adding support for additional space types and improving the efficiency of rendering for large grids.
- The PR includes updates to handle different sizes between the PropertyLayer and the grid space, issuing warnings when there's a mismatch.
- Colorbar display can be toggled using the `colorbar` parameter in the portrayal dictionary.
- The coordinate system has been validated to ensure consistent axis directions between PropertyLayers and the grid space.
- A demo using the Game of Life model has been implemented in a separate PR (https://github.com/projectmesa/mesa-examples/pull/214) to showcase these new visualization capabilities.

### TODO
- [x] Scale colors over alpha (aside from colormap)
- [x] Add switch for colorbars and fix warning
- [x] Validate coordinate system
  - [x] Handle different sizes (just warn)
  - [x] Axis in same direction
- [x] Demo GOL (https://github.com/projectmesa/mesa-examples/pull/214)
- [ ] Add tests
- [x] Write up PR

EwoutH commented:
This is ready for a first round of review, API wise its done. Any conceptual, feature or API feedback is welcome. If no major things come up, I will finish the tests and such.

The GoL example is also available here: https://github.com/projectmesa/mesa-examples/pull/214

EwoutH commented on mesa/visualization/components/matplotlib.py:
```diff
@@ -1,116 +1,177 @@
     update_counter.get()
     space_fig = Figure()
     space_ax = space_fig.subplots()
     space = getattr(model, "grid", None)
     if space is None:
```
This assumes the model either has a `space` or `grid`. I don't find this clean API design.

@Corvince, I would suggest allowing a `space` argument as input for the `make_space_matplotlib` and altair method. Then you can have as much spaces as you want and name them whatever you want.

rht replied:
`space` is the preferred naming. `grid` is there for legacy reason. The code here is so as to not break existing `model.grid` code.

quaquel replied:
Yes, but if you make explicit in the function signature what space/grid should be visualized, you can avoid the problem and give the user more flexibility.

EwoutH commented:
Implemented the last features (turning color bars on and off), validated the PropertyLayer coordinate system was visualized on the same axis as the Agent one, and updated the PR description.

![image](https://github.com/user-attachments/assets/be61e88e-b897-41c9-96cd-5e9c0925ac46)

Ready for review.

EwoutH commented:
Going to merge this, it breaks nothing and adds a new feature.

Post-merge review still appreciated!

rht commented:
> Going to merge this, it breaks nothing and adds a new feature.

This is not fine. You are breaking the convention of a review needed into a new feature *that is not experimental*. This is definitely not fair to other developers who have waited for a long time even for a review, e.g. @wang-boyu with the recent Mesa-Geo stuff, even though he is the lead maintainer on the repo.

EwoutH commented:
It’s a valid concern. I wanted to include it in the beta, so users could start testing it and delivering feedback, and updating example models (like https://github.com/projectmesa/mesa-examples/pull/214). I thought I had an approval on the API for this PR and #2341 from @quaquel.

@projectmesa/maintainers I would like your perspective on how to handle cases like this, for if they happen in the future.

rht commented on mesa/visualization/components/matplotlib.py:
```diff
@@ -1,116 +1,190 @@
+        space = getattr(model, "space", None)
+
+    if isinstance(space, mesa.space._Grid):
+        _draw_grid(space, space_ax, agent_portrayal, propertylayer_portrayal, model)
```
Either of:
- Why pass an entire model instead of just the layers?
- Why not use the space argument and access it from the model instead?

rht replied:
Edit: s/Why not use the space/Why not remove the space/

EwoutH replied:
Thanks for the review and the questions, I will try to look into them soon.

Also feel free to open a PR if you already have concrete solutions in mind.
